### PR TITLE
(SLV-556) Update perf_run_helper to process metrics collector and atop results

### PR DESCRIPTION
**Pull request by billclaytor**
# puppet metrics collector
This update adds functionality to perf_results_helper to process the puppet metrics collector data and extract the following:
* static_compile_mean
* average_borrow_time
* num_free_jrubies

The data is written to a CSV file with a row for each timestamp; the values are then averaged and written to a second CSV file. These files are converted to HTML for easy viewing.

# atop
With this update the atop CSV results are now automatically split into separate 'summary' and 'detail' files which previously required running a script after the performance run.

There has also been some additional refactoring of perf_results_helper. 

Note: spec tests for perf_results_helper are in progress; I will remove the DO NOT MERGE label once they're ready.

### CODE REVIEW

File: tests/helpers/perf_results_helper.rb [hunk 1]
```diff
@@ -120,21 +105,90 @@ def gatling2csv(results_dir)
...
+    if contents.nil? || contents.empty? || contents.keys.empty?
+      raise "The 'contents' element of the 'group' node must have at least one key"
+    end
+
+    # TODO: verify each key
+    # TODO: unit test to ensure data validity
+    puts "There are #{contents.keys.length} keys"
+    puts
+
+    (0..contents.keys.length - 1).each do |i|
```
Comment (johnduarte) [inline tests/helpers/perf_results_helper.rb]: Why is the last key skipped?

File: tests/helpers/perf_run_helper.rb [hunk 2]
```diff
@@ -23,6 +27,9 @@ module PerfRunHelper
   PERF_RESULTS_DIR = "results/perf"
   SCALE_RESULTS_DIR = "results/scale"
 
+  METRIC_RESULTS_DIR = "/root/gatling-puppet-load-test/simulation-runner/results"
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_run_helper.rb]: I made this a constant to allow re-use vs repetition:
https://github.com/puppetlabs/gatling-puppet-load-test/blob/739ed4edb32c245eb588c32c411843f1d26c2a23/tests/helpers/perf_run_helper.rb#L777

This is existing GPLT behavior. For example:
http://10.234.1.190/gplt/perf/slv/2019.1.0/PERF_1566923080/ip-10-227-3-214.amz-dev.puppet.net/root/

I don't believe running the simulation on the metrics node as a non-root user is supported.

File: tests/helpers/perf_run_helper.rb [hunk 3]
```diff
@@ -786,10 +807,40 @@ def get_perf_result
...
+  # Extracts the Gatling results data into a CSV file
+  #
+  # @author Bill Claytor
+  #
+  # @return [void]
+  #
+  # @example
+  #   process_gatling_json_file
+  #
+  def process_gatling_json_file
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_run_helper.rb]: Updated

File: tests/helpers/perf_run_helper.rb [hunk 4]
```diff
@@ -786,10 +807,40 @@ def get_perf_result
...
+  # Splits the atop CSV file into separate files for the summary and detail sections
+  #
+  # @author Bill Claytor
+  #
+  # @return [void]
+  #
+  # @example
+  #   process_atop_csv_file
+  #
+  def process_atop_csv_file
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_run_helper.rb]: Updated

File: tests/helpers/perf_results_helper.rb [hunk 5]
```diff
@@ -433,4 +593,130 @@ def percent_diff(result_a, result_b)
...
+    begin
+      timestamp = json["timestamp"]
+
+      # catalog
+      catalog_metrics = json["servers"][json["servers"].keys[0]]["puppetserver"]["pe-puppet-profiler"]["status"]["experimental"]["catalog-metrics"]
+
+      # ignore metrics without catalog metrics
+      # TODO: if this is an issue, investigate alternatives to handling averages
+
+      # TODO: remove
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: Updated

File: tests/helpers/perf_results_helper.rb [hunk 6]
```diff
@@ -433,4 +593,130 @@ def percent_diff(result_a, result_b)
...
+  #
+  # @param [String] file The JSON file to process
+  #
+  # @return [Array] The parameters array / csv row
+  #
+  # @example
+  #   process_puppetserver_json(file)
+  #
+  # rubocop:disable Metrics/LineLength
+  def process_puppetserver_json(file)
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: Updated

File: tests/helpers/perf_results_helper.rb [hunk 7]
```diff
@@ -433,4 +593,130 @@ def percent_diff(result_a, result_b)
...
+  #
+  # @param [String] metrics_dir The 'puppet-metrics-collector' directory
+  #
+  # @return [void]
+  #
+  # @example
+  #   process_puppetserver_files(metrics_dir)
+  #
+  # TODO: refactor to be generic
+  def process_puppetserver_files(metrics_dir)
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: Updated

File: tests/helpers/perf_results_helper.rb [hunk 8]
```diff
@@ -433,4 +593,130 @@ def percent_diff(result_a, result_b)
...
+  # @author Bill Claytor
+  #
+  # @param [String] metrics_dir The 'puppet-metrics-collector' directory
+  #
+  # @return [void]
+  #
+  # @example
+  #   extract_puppet_metrics_collector_data(metrics_dir)
+  #
+  def extract_puppet_metrics_collector_data(metrics_dir)
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: I used `extract_puppet_metrics_collector_data` as a wrapper to allow updating to add 'handlers' for the other service directories. `process_puppetserver_files` has been renamed `extract_puppetserver_metrics`.

File: tests/helpers/perf_results_helper.rb [hunk 9]
```diff
@@ -219,6 +275,106 @@ def csv2html(csv_path)
...
+        # only average if the total is > 0
+        average = if total.positive?
+                    total / num_rows
+                  else
+                    0
+                  end
+
+        averages_row << average.round(2)
+
+        # puts
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: Updated

File: tests/helpers/perf_results_helper.rb (hunk 1 above)
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: This is actually older code that was extracted. After adding tests to confirm that all keys are included I started to update it to just use the keys directly but realized that the intention was to reference the keys by index:
```
puts "key #{i}: #{name}")
```
I think this was initially added for debugging; not sure if it is still useful.

File: tests/helpers/perf_results_helper.rb [hunk 10]
```diff
@@ -433,4 +593,130 @@ def percent_diff(result_a, result_b)
...
+    row = nil
+
+    contents = File.read(file)
+    json = JSON.parse(contents)
+
+    begin
+      timestamp = json["timestamp"]
+
+      # catalog
+      catalog_metrics = json["servers"][json["servers"].keys[0]]["puppetserver"]["pe-puppet-profiler"]["status"]["experimental"]["catalog-metrics"]
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: Nice, thanks! I opened the following ticket to implement this suggestion:
https://tickets.puppetlabs.com/browse/SLV-569

File: tests/helpers/perf_results_helper.rb [hunk 11]
```diff
@@ -433,4 +593,130 @@ def percent_diff(result_a, result_b)
...
+      #
+      #   # jruby
+      #   pe_jruby_metrics = json["servers"][json["servers"].keys[0]]["puppetserver"]["pe-jruby-metrics"]["status"]["experimental"]["metrics"]
+      #   average_borrow_time = pe_jruby_metrics["average-borrow-time"]
+      #   num_free_jrubies = pe_jruby_metrics["num-free-jrubies"]
+      #   row = [timestamp, static_compile_mean, average_borrow_time, num_free_jrubies]
+      # end
+
+      # catalog
+      static_compile_mean = catalog_metrics[0]["mean"]
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: I added SLV-569 (https://tickets.puppetlabs.com/browse/SLV-569) to address this.

File: tests/helpers/perf_results_helper.rb [hunk 12]
```diff
@@ -433,4 +593,130 @@ def percent_diff(result_a, result_b)
...
+
+      # jruby
+      pe_jruby_metrics = json["servers"][json["servers"].keys[0]]["puppetserver"]["pe-jruby-metrics"]["status"]["experimental"]["metrics"]
```
[the remark this answers is not in the document]
Comment (billclaytor) [reply tests/helpers/perf_results_helper.rb]: See response regarding SLV-569 (https://tickets.puppetlabs.com/browse/SLV-569) above.

**Pull request merged**